File: src/main/kotlin/texlab/Range.kt
package texlab

import org.eclipse.lsp4j.Position
import org.eclipse.lsp4j.Range

fun range(startLine: Int,
          startCharacter: Int,
          endLine: Int,
          endCharacter: Int): Range {
    val start = Position(startLine, startCharacter)
    val end = Position(endLine, endCharacter)
    return Range(start, end)
}

fun Range.contains(position: Position): Boolean {
    fun leq(left: Position, right: Position): Boolean {
        return if (left.line == right.line) {
            left.character <= right.character
        } else {
            left.line <= right.line
        }
    }
    return leq(start, position) && leq(position, end)
}

fun Range.containsExclusive(position: Position): Boolean {
    fun lt(left: Position, right: Position): Boolean {
        return if (left.line == right.line) {
            left.character < right.character
        } else {
            left.line < right.line
        }
    }
    return lt(start, position) && lt(position, end)
}

File: src/range.ts
import { Position, Range } from 'vscode-languageserver';

export function contains(range: Range, position: Position): boolean {
  function leq(left: Position, right: Position) {
    return left.line === right.line
      ? left.character <= right.character
      : left.line <= right.line;
  }
  return leq(range.start, position) && leq(position, range.end);
}

export function containsExclusive(range: Range, position: Position): boolean {
  function lt(left: Position, right: Position) {
    return left.line === right.line
      ? left.character < right.character
      : left.line < right.line;
  }
  return lt(range.start, position) && lt(position, range.end);
}
